### PR TITLE
[com_content] - add articles ordering by votes, ratings

### PR DESCRIPTION
**Pull request by alikon**
Pull Request for Issue #10954 .
#### Summary of Changes

enable article ordering by votes or by rating on backend & frontend

articles oreder vote-rating

if vote plugin is enabled and show vote is setted to show

votes_medium

allow the module mod_articles_category to display articles with highest/lowest votes. or rating on frontend

modules articles category
#### Expected result

the articles are showed ordered by votes or by rating
modvote

### CODE REVIEW

**Comment by brianteeman** [comment]
I have tested this item :red_circle: unsuccessfully on 8fcbdf7405aae6d15c955e036414e32704eddde2<br /><br />I can see the column but I dont see the header saying Vote

Personally I see no value in this being enabled all the time. It should only ever be enabled if 
1. The vote plugin is enabled
and
2. The param "Show Voting" is set to show

Finally dont forget this would need to be enabled for the Hathor admin template
<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/10989">issues.joomla.org/joomla-cms/10989</a>.</sub>

**Comment by andrepereiradasilva** [comment]
agree with brian

also is the ordering by **number votes** or **vote rating**?

**Comment by alikon** [comment]
> It should only ever be enabled if 
1. The vote plugin is enabled
and
2. The param "Show Voting" is set to show

:thumbsup:  i'll do it next....

> also is the ordering by number votes or vote rating?

@andrepereiradasilva  should be easy to allow for both (as is,  is only "number of votes")

i'll do it next....
p.s.
not so sure  was an "interesting feature"

**Comment by brianteeman** [comment]
I still don't think it will be useful but if it's going to exist then it should be the best it can be

**Comment by andrepereiradasilva** [comment]
i think it's useful to allow to sort articles by number of votes or voting count.

**Comment by andrepereiradasilva** [comment]
i even think this could be expanded to category layouts "Article Order" ordering

**Comment by alikon** [comment]
please don't test  still work in progress

**Comment by alikon** [comment]
ok now is testable

**Comment by conconnl** [comment]
The idea is good, the things created are working.
I'm missing the Voting order in the Article Order options in the Menu items; Category Blog, Category List, Featured Articles and All Categories.
I'm missing the Show Voting option in Archived Articles menu item while it's available in Category Blog, Category List, Featured Articles and All Categories

Maybe I missed some other checks, but this is what I could find for now.<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/10989">issues.joomla.org/joomla-cms/10989</a>.</sub>

**Comment by alikon** [comment]
> The idea is good, the things created are working.
> I'm missing the Voting order in the Article Order options in the Menu items; Category Blog, Category List, Featured Articles and All Categories.
> I'm missing the Show Voting option in Archived Articles menu item while it's available in Category Blog, Category List, Featured Articles and All Categories

@conconnl should be ok now, can you check...

not so sure if i've catched them all :worried:

p.s
@brianteeman  sorry for the spam :stuck_out_tongue_winking_eye:

**Comment by conconnl** [comment]
@alikon I tested everything again.
I can't find any issues with the newly created options.

Maybe one more addition and then I think we got everything.

On the category list you can choose to display Author and Hits.
I think when you order by Votes you want to show the number of votes.
screen shot 2016-07-08 at 21 49 19

screen shot 2016-07-08 at 21 55 37

**Comment by alikon** [comment]
added the menu option
categorylist

and the frontend view
categorylistf

**Comment by brianteeman** [comment]
I have tested this item :red_circle: unsuccessfully on 05504b22d8fe799b2d3fb6d8ea9b0cea61fedc62<br /><br />1. In the admin it appears that the default for Show Votes in List and Show Ratings in List in the menu item is Show. It must be Use Global or it changes the frontend of an existing site on upgrade. Also the default value in the component option must be hide for the same reason

2. In the front end I am getting 
PHP Parse error:  syntax error, unexpected '?' in /****/components/com_content/views/category/tmpl/default_articles.php on line 223<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/10989">issues.joomla.org/joomla-cms/10989</a>.</sub>

**Comment by brianteeman** [comment]
Also it appears the latest PR loses the ability to see the list of votes etc in the admin but I can still sort - is that correct?<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/10989">issues.joomla.org/joomla-cms/10989</a>.</sub>

**Comment by alikon** [comment]
>  1. default

fixed



> loses the ability to see the list of votes etc in the admin but I can still sort - is that correct?

dont understand , can you elaborate....

**Comment by brianteeman** [comment]
See screenshot of admin
zqxd

**Comment by alikon** [comment]
Vote/Rating column appear only when you select  one of these: 

- vote descending
- vote ascending
- rating descending
- rating ascending

votessel

**Comment by brianteeman** [comment]
Tried that - nothing showed up and then i got the following error

> You have an error in your SQL syntax; check the manual that corresponds
to your MySQL server version for the right syntax to use near 'LIMIT 20' at
line 13 SQL=SELECT a.id, a.title, a.alias, a.checked_out,
a.checked_out_time, a.catid, a.state, a.access, a.created, a.created_by,
a.created_by_alias, a.ordering, a.featured, a.language, a.hits,
a.publish_up, a.publish_down,l.title AS language_title, l.image AS
language_image,uc.name AS editor,ag.title AS access_level,c.title AS
category_title,ua.name AS author_name,COALESCE(NULLIF(ROUND(v.rating_sum /
v.rating_count, 0), 0), 0) AS rating, COALESCE(NULLIF(v.rating_count, 0),
0) as rating_count,COUNT(asso2.id)>1 as association FROM #__content AS a
LEFT JOIN `#__languages` AS l ON l.lang_code = a.language LEFT JOIN
#__users AS uc ON uc.id=a.checked_out LEFT JOIN #__viewlevels AS ag ON ag.id
= a.access LEFT JOIN #__categories AS c ON c.id = a.catid LEFT JOIN
#__users AS ua ON ua.id = a.created_by LEFT JOIN #__content_rating AS v ON
a.id = v.content_id LEFT JOIN #__associations AS asso ON asso.id = a.id AND
asso.context='com_content.item' LEFT JOIN #__associations AS asso2 ON
asso2.key = asso.key WHERE (a.state = 0 OR a.state = 1) GROUP BY a.id,
l.title, l.image, uc.name, ag.title, c.title, ua.name ORDER BY LIMIT 20 You
have an error in your SQL syntax; check the manual that corresponds to your
MySQL server version for the right syntax to use near '' at line 13
SQL=SELECT a.id, a.title, a.alias, a.checked_out, a.checked_out_time,
a.catid, a.state, a.access, a.created, a.created_by, a.created_by_alias,
a.ordering, a.featured, a.language, a.hits, a.publish_up,
a.publish_down,l.title AS language_title, l.image AS language_image,uc.name
AS editor,ag.title AS access_level,c.title AS category_title,ua.name AS
author_name,COALESCE(NULLIF(ROUND(v.rating_sum / v.rating_count, 0), 0), 0)
AS rating, COALESCE(NULLIF(v.rating_count, 0), 0) as rating_count,COUNT(
asso2.id)>1 as association FROM #__content AS a LEFT JOIN `#__languages` AS
l ON l.lang_code = a.language LEFT JOIN #__users AS uc ON uc.id=a.checked_out
LEFT JOIN #__viewlevels AS ag ON ag.id = a.access LEFT JOIN #__categories
AS c ON c.id = a.catid LEFT JOIN #__users AS ua ON ua.id = a.created_by
LEFT JOIN #__content_rating AS v ON a.id = v.content_id LEFT JOIN
#__associations AS asso ON asso.id = a.id AND
asso.context='com_content.item' LEFT JOIN #__associations AS asso2 ON
asso2.key = asso.key WHERE (a.state = 0 OR a.state = 1) GROUP BY a.id,
l.title, l.image, uc.name, ag.title, c.title, ua.name ORDER BY


Going out now for the day


On 9 July 2016 at 09:19, Nicola Galgano <notifications@github.com> wrote:

> Vote/Rating column appear only when you select one of these:
>
>    - vote descending
>    - vote ascending
>    - rating descending
>    - rating ascending
>
> [image: votessel]
> <https://cloud.githubusercontent.com/assets/181681/16706915/80bb29cc-45be-11e6-816b-e4a7b6630511.PNG>
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/joomla/joomla-cms/pull/10989#issuecomment-231522749>,
> or mute the thread
> <https://github.com/notifications/unsubscribe/ABPH8WbWB6-nQY7dsSBTgEvbwpFWrxauks5qT1mTgaJpZM4JDVHd>
> .
>



-- 
Brian Teeman
Co-founder Joomla! and OpenSourceMatters Inc.
http://brian.teeman.net/

**Comment by alikon** [comment]
fixed when multilaguage on, sorry

**Comment by brianteeman** [comment]
Re-applied patch but no columns in the admin

screen shot 2016-07-10 at 05 09 37<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/10989">issues.joomla.org/joomla-cms/10989</a>.</sub>

File: administrator/components/com_content/models/forms/filter_articles.xml [hunk 1]
```diff
@@ -114,6 +114,10 @@
 			<option value="a.featured DESC">JFEATURED_DESC</option>
 			<option value="a.hits ASC">JGLOBAL_HITS_ASC</option>
 			<option value="a.hits DESC">JGLOBAL_HITS_DESC</option>
+			<option value="rating_count ASC">JGLOBAL_VOTES_ASC</option>
+			<option value="rating_count DESC">JGLOBAL_VOTES_DESC</option>
+			<option value="rating ASC">JGLOBAL_RATINGS_ASC</option>
+			<option value="rating DESC">JGLOBAL_RATINGS_DESC</option>
```
Comment (andrepereiradasilva) [inline administrator/components/com_content/models/forms/filter_articles.xml]: IMHO this 4 options should only appear when content vote plugin is enabled.

I think you can follow the same logic as associations or multilanguage ex: `requires="voting"`

See https://github.com/joomla/joomla-cms/blob/staging/libraries/joomla/form/fields/list.php#L107-L120

File: administrator/components/com_content/models/articles.php [hunk 2]
```diff
@@ -310,26 +314,10 @@ protected function getListQuery()
 		}
 
 		// Add the list ordering clause.
-		$orderCol = $this->state->get('list.ordering', 'a.id');
+		$orderCol = $this->state->get('list.fullordering', 'a.id');
```
Comment (andrepereiradasilva) [inline administrator/components/com_content/models/articles.php]: why did you change it to fullordering?

i think this breaks the ordering in list view (you can see there is no order arrow in the column titles)

File: administrator/components/com_content/views/articles/tmpl/default.php [hunk 3]
```diff
@@ -18,7 +18,7 @@
 $app       = JFactory::getApplication();
 $user      = JFactory::getUser();
 $userId    = $user->get('id');
-$listOrder = $this->escape($this->state->get('list.ordering'));
+$listOrder = $this->escape($this->state->get('list.fullordering', 'a.id'));
```
Comment (andrepereiradasilva) [inline administrator/components/com_content/views/articles/tmpl/default.php]: same comment as above

File: administrator/components/com_content/views/articles/tmpl/default.php [hunk 4]
```diff
@@ -29,6 +29,21 @@
...
+	$vote = true;
+}
+
+$votes  = ($listOrder == 'rating_count DESC')||($listOrder == 'rating_count ASC');
+$rating = ($listOrder == 'rating DESC')||($listOrder == 'rating ASC');
+
+if (!(($votes)||($rating)))
+{
+	$vote = false;
+}
```
Comment (andrepereiradasilva) [inline administrator/components/com_content/views/articles/tmpl/default.php]: IMHO this checks should be in the view file, not in the layout file.

File: administrator/modules/mod_latest/helper.php [hunk 5]
```diff
@@ -40,13 +40,13 @@ public static function getList(&$params)
 		switch ($params->get('ordering'))
 		{
 			case 'm_dsc':
-				$model->setState('list.ordering', 'modified DESC, created');
+				$model->setState('list.fullordering', 'modified DESC, created');
```
Comment (andrepereiradasilva) [inline administrator/modules/mod_latest/helper.php]: same comment as above

File: administrator/modules/mod_latest/helper.php [hunk 6]
```diff
@@ -40,13 +40,13 @@ public static function getList(&$params)
...
 				$model->setState('list.direction', 'DESC');
 				break;
 
 			case 'c_dsc':
 			default:
-				$model->setState('list.ordering', 'created');
+				$model->setState('list.fullordering', 'created');
```
Comment (andrepereiradasilva) [inline administrator/modules/mod_latest/helper.php]: same comment as above

File: administrator/modules/mod_popular/helper.php [hunk 7]
```diff
@@ -37,7 +37,7 @@ public static function getList(&$params)
 				' a.created, a.hits');
 
 		// Set Ordering filter
-		$model->setState('list.ordering', 'a.hits');
+		$model->setState('list.fullordering', 'a.hits DESC');
```
Comment (andrepereiradasilva) [inline administrator/modules/mod_popular/helper.php]: same comment as above

File: administrator/components/com_content/models/forms/filter_articles.xml (hunk 1 above)
Comment (alikon) [inline administrator/components/com_content/models/forms/filter_articles.xml]: good tip 

now only appear when content vote plugin is enabled.

File: administrator/components/com_content/models/articles.php (hunk 2 above)
Comment (alikon) [inline administrator/components/com_content/models/articles.php]: cause of https://github.com/joomla/joomla-cms/blob/staging/administrator/components/com_content/models/forms/filter_articles.xml#L85
i've used the old one on layout

File: administrator/components/com_content/views/articles/tmpl/default.php (hunk 4 above)
Comment (alikon) [inline administrator/components/com_content/views/articles/tmpl/default.php]: yes agree

**Comment by alikon** [comment]
@brianteeman 
now the vote/rate sort options only appears when the vote plugin is enabled as suggested by @andrepereiradasilva

**Comment by brianteeman** [comment]
1. I see now that you only get the option to sort by vote/rating when the plugin is enabled. Cant say i am a fan though that the vote and/or rating are only displayed when you sort by those options. We dont do that for hits for example. And you can only see one at a time

2. If you have an existing menu item set to display a list of articles and apply this patch then the votes and ratings are displayed. You have to go and open and save the menu item for them not to be displayed. This will be a real pain on a site with hundreds of menu items.

3. Does it make sense that you can set the menu item for a list of articles to display the rating and votes if the plugin is not enabled? I dont think it does.<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/10989">issues.joomla.org/joomla-cms/10989</a>.</sub>

**Comment by alikon** [comment]
>  1, 2, 3. 

 (make sense)

- should look like this if the plugin is enabled (the 2 coloumns are displayed even if the vote/rate order is not selected)

pluginenabled

- and the options settings  are enabled and setted to the default hide


pluginenabled2

- with the vote plugin  disabled (the 2 coloumns are not displayed  nor the vote/rate order sort options)

plugindisabled

- and the options settings  are unabled 

plugindisabled2

File: administrator/components/com_content/views/articles/view.html.php [hunk 8]
```diff
@@ -42,6 +42,13 @@ public function display($tpl = null)
 		$this->authors       = $this->get('Authors');
 		$this->filterForm    = $this->get('FilterForm');
 		$this->activeFilters = $this->get('ActiveFilters');
+		$this->vote          = false;
+		$listOrder           = $this->state->get('list.fullordering', 'a.id');
+
+		if (JPluginHelper::isEnabled('content', 'vote'))
+		{
+			$this->vote = true;
+		}
```
Comment (andrepereiradasilva) [inline administrator/components/com_content/views/articles/view.html.php]: micro optimization
JPluginHelper returns a boolean value (see https://github.com/joomla-projects/gsoc16_improved-multi-lingual/blob/staging/libraries/cms/plugin/helper.php#L126) so this would be enough.

```php
$this->vote          = JPluginHelper::isEnabled('content', 'vote')
```;

File: components/com_content/views/category/tmpl/default.xml [hunk 9]
```diff
@@ -227,6 +251,10 @@
 				<option value="rhits">JGLOBAL_LEAST_HITS</option>
 				<option value="random">JGLOBAL_RANDOM_ORDER</option>
 				<option value="order">JGLOBAL_ORDERING</option>
+				<option value="vote">JGLOBAL_VOTES_DESC</option>
+				<option value="rvote">JGLOBAL_VOTES_ASC</option>
+				<option value="rank">JGLOBAL_RATINGS_DESC</option>
+				<option value="rrank">JGLOBAL_RATINGS_ASC</option>
```
Comment (andrepereiradasilva) [inline components/com_content/views/category/tmpl/default.xml]: `requires="vote"` ?

File: administrator/components/com_content/views/articles/view.html.php (hunk 8 above)
Comment (alikon) [inline administrator/components/com_content/views/articles/view.html.php]: yes, micro optimization done
thanks

File: components/com_content/views/featured/tmpl/default.xml [hunk 10]
```diff
@@ -91,6 +91,10 @@
 				<option value="hits">JGLOBAL_MOST_HITS</option>
 				<option value="rhits">JGLOBAL_LEAST_HITS</option>
 				<option value="order">JGLOBAL_ORDERING</option>
+				<option value="vote">JGLOBAL_VOTES_DESC</option>
+				<option value="rvote">JGLOBAL_VOTES_ASC</option>
+				<option value="rank">JGLOBAL_RATINGS_DESC</option>
+				<option value="rrank">JGLOBAL_RATINGS_ASC</option>
```
Comment (andrepereiradasilva) [inline components/com_content/views/featured/tmpl/default.xml]: `requires="vote"` ?

File: modules/mod_articles_category/mod_articles_category.xml [hunk 11]
```diff
@@ -295,6 +295,8 @@
 					<option value="publish_up">MOD_ARTICLES_CATEGORY_OPTION_STARTPUBLISHING_VALUE</option>
 					<option value="a.publish_down">MOD_ARTICLES_CATEGORY_OPTION_FINISHPUBLISHING_VALUE</option>
 					<option value="random">MOD_ARTICLES_CATEGORY_OPTION_RANDOM_VALUE</option>
+					<option value="rating_count">MOD_ARTICLES_CATEGORY_OPTION_VOTE_VALUE</option>
+					<option value="rating">MOD_ARTICLES_CATEGORY_OPTION_RATING_VALUE</option>
```
Comment (andrepereiradasilva) [inline modules/mod_articles_category/mod_articles_category.xml]: `requires="vote"`?

File: components/com_content/views/category/tmpl/blog.xml [hunk 12]
```diff
@@ -264,6 +264,10 @@
 					<option value="rhits">JGLOBAL_LEAST_HITS</option>
 					<option value="random">JGLOBAL_RANDOM_ORDER</option>
 					<option value="order">JGLOBAL_ORDERING</option>
+					<option value="vote">JGLOBAL_VOTES_DESC</option>
+					<option value="rvote">JGLOBAL_VOTES_ASC</option>
+					<option value="rank">JGLOBAL_RATINGS_DESC</option>
+					<option value="rrank">JGLOBAL_RATINGS_ASC</option>
```
Comment (andrepereiradasilva) [inline components/com_content/views/category/tmpl/blog.xml]: `requires="vote"` ?

File: components/com_content/views/categories/tmpl/default.xml [hunk 13]
```diff
@@ -261,6 +261,10 @@
 				<option value="hits">JGLOBAL_MOST_HITS</option>
 				<option value="rhits">JGLOBAL_LEAST_HITS</option>
 				<option value="order">JGLOBAL_ORDERING</option>
+				<option value="vote">JGLOBAL_VOTES_DESC</option>
+				<option value="rvote">JGLOBAL_VOTES_ASC</option>
+				<option value="rank">JGLOBAL_RATINGS_DESC</option>
+				<option value="rrank">JGLOBAL_RATINGS_ASC</option>
```
Comment (andrepereiradasilva) [inline components/com_content/views/categories/tmpl/default.xml]: `requires="vote"` ?

File: components/com_content/views/archive/tmpl/default.xml [hunk 14]
```diff
@@ -30,6 +30,10 @@
 				<option value="hits">JGLOBAL_MOST_HITS</option>
 				<option value="rhits">JGLOBAL_LEAST_HITS</option>
 				<option value="order">JGLOBAL_ARTICLE_MANAGER_ORDER</option>
+				<option value="vote">JGLOBAL_VOTES_DESC</option>
+				<option value="rvote">JGLOBAL_VOTES_ASC</option>
+				<option value="rank">JGLOBAL_RATINGS_DESC</option>
+				<option value="rrank">JGLOBAL_RATINGS_ASC</option>
```
Comment (andrepereiradasilva) [inline components/com_content/views/archive/tmpl/default.xml]: `requires="vote"` ?

File: components/com_content/models/articles.php [hunk 15]
```diff
@@ -253,7 +253,7 @@ protected function getListQuery()
 			->join('LEFT', '#__categories as parent ON parent.id = c.parent_id');
 
 		// Join on voting table
-		$query->select('ROUND(v.rating_sum / v.rating_count, 0) AS rating, v.rating_count as rating_count')
+		$query->select('COALESCE(NULLIF(ROUND(v.rating_sum  / v.rating_count, 0), 0), 0) AS rating, COALESCE(NULLIF(v.rating_count, 0), 0) as rating_count')
 			->join('LEFT', '#__content_rating AS v ON a.id = v.content_id');
```
Comment (andrepereiradasilva) [inline components/com_content/models/articles.php]: for performance reasons shouldn't this be done only if vote pluign is enabled and component param list/order selected?

File: administrator/components/com_content/views/articles/view.html.php [hunk 16]
```diff
@@ -42,6 +42,8 @@ public function display($tpl = null)
 		$this->authors       = $this->get('Authors');
 		$this->filterForm    = $this->get('FilterForm');
 		$this->activeFilters = $this->get('ActiveFilters');
+		$this->vote          = JPluginHelper::isEnabled('content', 'vote');
+		$listOrder           = $this->state->get('list.fullordering', 'a.id');
```
Comment (andrepereiradasilva) [inline administrator/components/com_content/views/articles/view.html.php]: is this really neeeded?

File: administrator/components/com_content/models/articles.php [hunk 17]
```diff
@@ -199,13 +199,17 @@ protected function getListQuery()
 		$query->select('ua.name AS author_name')
 			->join('LEFT', '#__users AS ua ON ua.id = a.created_by');
 
+		// Join on voting table
+		$query->select('COALESCE(NULLIF(ROUND(v.rating_sum  / v.rating_count, 0), 0), 0) AS rating, COALESCE(NULLIF(v.rating_count, 0), 0) as rating_count')
+			->join('LEFT', '#__content_rating AS v ON a.id = v.content_id');
```
Comment (andrepereiradasilva) [inline administrator/components/com_content/models/articles.php]: for performance reasons shouldn't this be done only if vote pluign is enabled and component param list/order selected?

**Comment by alikon** [comment]
closed for conflicts in favour of #11225
